### PR TITLE
Roles management

### PR DESCRIPTION
**Pull request by arhimondr**
Cherry picked from

- https://github.com/prestosql/presto/pull/90 (originally https://github.com/prestodb/presto/pull/11645)
- https://github.com/prestosql/presto/pull/113
- https://github.com/prestosql/presto/pull/291

### CODE REVIEW

**Comment by wenleix** [comment]
Why https://github.com/prestosql/presto/pull/90 adds ~8000 lines while this PR only adds ~7800 lines ? :P

**Comment by arhimondr** [comment]
>     * `TestSqlParserErrorHandling.java` has slightly different change.

For some reason a test was duplicated in prestosql

**Comment by arhimondr** [comment]
> Why [prestosql/presto#90](https://github.com/prestosql/presto/pull/90) adds ~8000 lines while this PR only adds ~7800 lines ? :P

Fixed

**Comment by arhimondr** [comment]
> Why [prestosql/presto#90](https://github.com/prestosql/presto/pull/90) adds ~8000 lines while this PR only adds ~7800 lines ? :P

I picked that from the https://github.com/prestodb/presto/pull/11645, as literally cherry picking of 65 commits from the prestosql is too much of work. It is curious where the 150 additions were lost. We have 100 more removals. Maybe some difference because of rebasing / repackaging?

**Comment by wenleix** [comment]
Looks like file `ClientSelectedRole.java` is missing from Implement SET ROLE (compared with https://github.com/prestosql/presto/pull/90/commits/e02bdabfe6a7bea3459b4d3147f36fe85f3554d3)

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoConnection.java [hunk 1]
```diff
@@ -16,6 +16,8 @@
 import com.facebook.presto.client.ClientSession;
 import com.facebook.presto.client.ServerInfo;
 import com.facebook.presto.client.StatementClient;
+import com.facebook.presto.spi.security.SelectedRole;
```
Comment (wenleix) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoConnection.java]: The change in this file is slightly different from https://github.com/prestosql/presto/pull/90/commits/e02bdabfe6a7bea3459b4d3147f36fe85f3554d3

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java [hunk 2]
```diff
@@ -248,6 +249,10 @@ final boolean internalExecute(String sql)
             currentWarningsManager.set(Optional.of(warningsManager));
             resultSet = new PrestoResultSet(client, maxRows.get(), progressConsumer, warningsManager);
 
+            for (Map.Entry<String, SelectedRole> entry : client.getSetRoles().entrySet()) {
```
Comment (wenleix) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: Also this file.

**Comment by arhimondr** [comment]
@wenleix 

> For "Implement Create/Drop/List roles in Hive connector": It looks like changes in `TestRecordingHiveMetastore.java` is missing (compared with [prestosql/presto@22de2d8](https://github.com/prestosql/presto/commit/22de2d81a995cb7687953413e5048e5a95484931))

This change is missing in that commit in prestosql. Probably while rebasing the prestosql version they didn't compile every commit.

**Comment by arhimondr** [comment]
@wenleix 

> Looks like file `ClientSelectedRole.java` is missing from Implement SET ROLE (compared with [prestosql/presto@e02bdab](https://github.com/prestosql/presto/commit/e02bdabfe6a7bea3459b4d3147f36fe85f3554d3))

In prestosql they removed SPI from the `presto-client` dependencies. That's why they had to copy the `SelectedRole` class to the `presto-client` package.

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoConnection.java (hunk 1 above)
Comment (arhimondr) [reply presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoConnection.java]: https://github.com/prestodb/presto/pull/12383#issuecomment-467633410

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java (hunk 2 above)
Comment (arhimondr) [reply presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: https://github.com/prestodb/presto/pull/12383#issuecomment-467633410

File: presto-main/src/main/java/com/facebook/presto/Session.java [hunk 3]
```diff
@@ -327,11 +327,9 @@ public Session beginTransactionId(TransactionId transactionId, TransactionManage
         for (Entry<String, SelectedRole> entry : identity.getRoles().entrySet()) {
             String catalogName = entry.getKey();
             SelectedRole role = entry.getValue();
-
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/Session.java]: This commit looks does nothing . But let's keep consistent with PrestoSQL ;)

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/Session.java]: What commit is that?

File: presto-main/src/main/java/com/facebook/presto/connector/ConnectorManager.java [hunk 4]
```diff
@@ -196,7 +196,7 @@ private synchronized void addCatalogConnector(String catalogName, ConnectorId co
 
         MaterializedConnector informationSchemaConnector = new MaterializedConnector(
                 createInformationSchemaConnectorId(connectorId),
-                new InformationSchemaConnector(catalogName, nodeManager, metadataManager, accessControlManager));
+                new InformationSchemaConnector(catalogName, nodeManager, metadataManager, accessControlManager, transactionManager));
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/connector/ConnectorManager.java]: This change and `InformationSchemaConnector.java` doesn't appear in https://github.com/prestosql/presto/pull/90/commits/7baabd1f658a6eccfeec61101f61cbaf3c972d13 :). Maybe again get dropped on PrestoSQL side.

File: presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java [hunk 5]
```diff
@@ -66,11 +68,13 @@
 {
     private final Metadata metadata;
     private final AccessControl accessControl;
+    private final TransactionManager transactionManager;
 
-    public InformationSchemaPageSourceProvider(Metadata metadata, AccessControl accessControl)
+    public InformationSchemaPageSourceProvider(Metadata metadata, AccessControl accessControl, TransactionManager transactionManager)
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java]: This change does't appear in https://github.com/prestosql/presto/pull/90/commits/7baabd1f658a6eccfeec61101f61cbaf3c972d13 .  -- also doesn't seem useful?

File: presto-main/src/test/java/com/facebook/presto/security/TestAccessControlManager.java [hunk 6]
```diff
@@ -198,7 +198,7 @@ private static ConnectorId registerBogusConnector(CatalogManager catalogManager,
                 connectorId,
                 connector,
                 createInformationSchemaConnectorId(connectorId),
-                new InformationSchemaConnector(catalogName, nodeManager, metadata, accessControl),
+                new InformationSchemaConnector(catalogName, nodeManager, metadata, accessControl, transactionManager),
```
Comment (wenleix) [inline presto-main/src/test/java/com/facebook/presto/security/TestAccessControlManager.java]: Looks like https://github.com/prestosql/presto/pull/90/commits/7baabd1f658a6eccfeec61101f61cbaf3c972d13 removed all the `transactionManager` related stuff.

File: presto-main/src/main/java/com/facebook/presto/security/AllowAllSystemAccessControl.java [hunk 7]
```diff
@@ -187,4 +187,15 @@ public void checkCanGrantTablePrivilege(Identity identity, Privilege privilege,
...
     {
     }
+
+    @Override
+    public void checkCanShowRoles(Identity identity, String catalogName)
+    {
+    }
+
+    @Override
+    public Set<String> filterRoles(Identity identity, String catalogName, Set<String> roles)
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/security/AllowAllSystemAccessControl.java]: This method was dropped in https://github.com/prestosql/presto/pull/90/commits/e2a2ff03d78734065ef0165225cebbb9dc3af89f

File: presto-spi/src/main/java/com/facebook/presto/spi/security/SystemAccessControl.java [hunk 8]
```diff
@@ -301,4 +302,22 @@ default void checkCanRevokeTablePrivilege(Identity identity, Privilege privilege
...
+     */
+    default void checkCanShowRoles(Identity identity, String catalogName)
+    {
+        denyShowRoles(catalogName);
+    }
+
+    /**
+     * Filter the list of roles to those visible to the identity in the given catalog.
+     */
+    default Set<String> filterRoles(Identity identity, String catalogName, Set<String> roles)
```
Comment (wenleix) [inline presto-spi/src/main/java/com/facebook/presto/spi/security/SystemAccessControl.java]: dropped in https://github.com/prestosql/presto/pull/90/commits/e2a2ff03d78734065ef0165225cebbb9dc3af89f

File: presto-main/src/main/java/com/facebook/presto/security/AllowAllSystemAccessControl.java [hunk 9]
```diff
@@ -187,15 +187,4 @@ public void checkCanGrantTablePrivilege(Identity identity, Privilege privilege,
     public void checkCanRevokeTablePrivilege(Identity identity, Privilege privilege, CatalogSchemaTableName table, PrestoPrincipal revokee, boolean grantOptionFor)
     {
     }
-
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/security/AllowAllSystemAccessControl.java]: Doesn't see this change in https://github.com/prestosql/presto/pull/90/commits/4f4b268adce8a214b73344a82c872bd493ef55ac . Maybe they already implement in different commit?

File: presto-main/src/main/java/com/facebook/presto/security/ReadOnlySystemAccessControl.java [hunk 10]
```diff
@@ -110,15 +110,4 @@ public void checkCanShowSchemas(Identity identity, String catalogName)
     public void checkCanShowTablesMetadata(Identity identity, CatalogSchemaName schema)
     {
     }
-
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/security/ReadOnlySystemAccessControl.java]: same for this one.

File: presto-plugin-toolkit/src/main/java/com/facebook/presto/plugin/base/security/ForwardingConnectorAccessControl.java [hunk 11]
```diff
@@ -205,4 +205,22 @@ public void checkCanSetRole(ConnectorTransactionHandle transactionHandle, Connec
     {
         delegate().checkCanSetRole(transactionHandle, identity, role, catalogName);
     }
+
+    @Override
+    public void checkCanShowRoles(ConnectorTransactionHandle transactionHandle, ConnectorIdentity identity, String catalogName)
```
Comment (wenleix) [inline presto-plugin-toolkit/src/main/java/com/facebook/presto/plugin/base/security/ForwardingConnectorAccessControl.java]: https://github.com/prestosql/presto/pull/90/commits/4f4b268adce8a214b73344a82c872bd493ef55ac#diff-cc0689ca452d7c2209c27e411d80c45aR177 also adds `checkCanSetRole`

File: presto-spi/src/main/java/com/facebook/presto/spi/security/SystemAccessControl.java [hunk 12]
```diff
@@ -302,22 +301,4 @@ default void checkCanRevokeTablePrivilege(Identity identity, Privilege privilege
     {
         denyRevokeTablePrivilege(privilege.toString(), table.toString());
     }
-
-    /**
```
Comment (wenleix) [inline presto-spi/src/main/java/com/facebook/presto/spi/security/SystemAccessControl.java]: same for this one.

**Comment by wenleix** [comment]
Finish "human diff" the following commits:


* Introduce GRANT/REVOKE roles statements
* Add Grant/Revoke/List roles authorization to the SPI 
* Introduce APPLICABLE_ROLES view 
* Implement Grant/Revoke/ListApplicableRoles in Hive 
* Refactor GRANT/REVOKE in Hive
* Introduce access control for GRANT/REVOKE ROLE
* Prepare metastore interface to accept ROLE for GRANT/REVOKE
* Introduce SET ROLE statement
* Introduce ConnectorIdentity 
* Implement SET ROLE
* Make HiveQueryRunner default user have admin role
* Set admin role before setting system properties in storage format test
* Store catalog selected roles in Identity
* Implement SET ROLE in Hive Connector
* Accept ROLE in GRANT/REVOKE Privileges statements
* Add SHOW ROLES to the parser
* Add access control checks for SHOW ROLES
* Add docs for SHOW ROLES
* Product tests for SHOW ROLES
* Add SHOW CURRENT ROLES
* Add SHOW ROLE GRANTS syntax
* Add listRoleGrants to the SPI 
* Implement listRoleGrants() in Hive 
* Implement SHOW ROLE GRANTS rewrite
* Add docs for SHOW ROLE GRANTS
* Access control for SHOW ROLE GRANTS and SHOW CURRENT ROLES

File: presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestGrantRevoke.java [hunk 13]
```diff
@@ -72,12 +79,32 @@ public void cleanup()
...
             aliceExecutor.executeQuery(format("DROP TABLE IF EXISTS %s", tableName));
             aliceExecutor.executeQuery(format("DROP VIEW IF EXISTS %s", viewName));
+            cleanupRoles();
         }
         catch (Exception e) {
             Logger.get(getClass()).warn(e, "failed to drop table/view");
         }
     }
 
+    private void cleanupRoles()
```
Comment (wenleix) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestGrantRevoke.java]: https://github.com/prestosql/presto/pull/90/commits/6735d23b22ca17538000cabde4f3d183e894432a adds annotation `@Test(groups = {AUTHORIZATION, PROFILE_SPECIFIC_TESTS})`

File: presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java [hunk 14]
```diff
@@ -563,6 +564,119 @@ public void testShowRoleGrants()
                         row("role1"));
     }
 
+    @Test(groups = {ROLES, AUTHORIZATION, PROFILE_SPECIFIC_TESTS})
```
Comment (wenleix) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java]: https://github.com/prestosql/presto/pull/90/commits/aec0671f2956200c6e562435a601edaad53ce3e3 also renamed `testAccessControl` to `testCreateDropRoleAccessControl` . Although seems unrelated.

File: presto-main/src/main/java/com/facebook/presto/Session.java (hunk 3 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/Session.java]: Found it: `Store catalog selected roles in Identity`. Let me drop it.

**Comment by wenleix** [comment]
Finish "human diff" the following commits against https://github.com/prestosql/presto/pull/90/

* Remove redundant checkDatabasePermission methods 
* Reorder methods in SqlStandardAccessControl
* Rename getGrantOptionForPrivilege to hasGrantOptionForPrivilege
* Remove hive privilege null check
* Allow all for admin role
* Introduce isTableOwner method for readability
* Simplify checkTablePermission
* Refactor canCreateView security checks
* Consider enabled roles for permissions
* Refactor HivePrivilegeInfo  
* Move parsePrivilege to MetastoreUtil
* Add grantor to HivePrivilegeInfo
* Add grantor_type and grantee_type columns to table_privileges
* More product tests for SET ROLE
* Document role management

File: presto-main/src/main/java/com/facebook/presto/Session.java (hunk 3 above)
Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/Session.java]: I see. Sorry for forgetting to use "Add Review" mode...

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreUtil.java [hunk 15]
```diff
@@ -218,60 +221,66 @@ public static PrivilegeGrantInfo toMetastoreApiPrivilegeGrantInfo(HivePrivilegeI
...
-        while (!queue.isEmpty()) {
-            PrestoPrincipal current = queue.poll();
-            Set<RoleGrant> grants = listRoleGrants.apply(current);
-            for (RoleGrant grant : grants) {
-                if (!result.contains(grant)) {
-                    result.add(grant);
-                    queue.add(new PrestoPrincipal(ROLE, grant.getRoleName()));
+        Queue<RoleGrant> output = new ArrayDeque<>();
+        Set<RoleGrant> seenRoles = new HashSet<>();
+        return Streams.stream(new AbstractIterator<RoleGrant>() {
```
Comment (wenleix) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreUtil.java]: hmm... is the convention to have a new line or not? (diff from https://github.com/prestosql/presto/pull/90/commits/7b73ba0aead016f3b37a1df30b98905a30f4d799)

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/UserTableKey.java [hunk 16]
```diff
@@ -37,7 +37,7 @@ public UserTableKey(
             @JsonProperty("database") String database,
             @JsonProperty("table") String table)
     {
-        this.principal = requireNonNull(principal, "principal is null");
+        this.principal = principal;
```
Comment (wenleix) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/UserTableKey.java]: there is a comment in https://github.com/prestosql/presto/pull/90/commits/0cf6e5b0a3a0365e4bcfad6d0413fc1a7b8cdb65:
```
        // principal can be null when we want to list all privileges for admins
```

**Comment by wenleix** [comment]
Finish "human diff" the following commits against [prestosql/presto#90](https://github.com/prestosql/presto/pull/90)

* Remove unused metastore statistics beans
* Extract role name constants
* Enumerate roles until enabled role is found 
* Make ThriftMetastoreUtil.list*Roles methods to return Stream 
* Make list*TablePrivileges to return Stream
* Fix listing privileges of tables owned by an user 
* Fix listing privileges when admin role is set
* Fix rewrite SHOW GRANTS as a SELECT query
* Remove 'IN catalog' syntax from role management commands
* Refactor RecordingHiveMetastore to make it consistent between releases

File: presto-main/src/main/java/com/facebook/presto/connector/ConnectorManager.java (hunk 4 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/connector/ConnectorManager.java]: Yeah, good catch

File: presto-main/src/main/java/com/facebook/presto/security/AllowAllSystemAccessControl.java (hunk 9 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/security/AllowAllSystemAccessControl.java]: Recovered in `Add access control checks for SHOW ROLES`

File: presto-main/src/main/java/com/facebook/presto/security/ReadOnlySystemAccessControl.java (hunk 10 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/security/ReadOnlySystemAccessControl.java]: Recovered in `Add access control checks for SHOW ROLES`

File: presto-plugin-toolkit/src/main/java/com/facebook/presto/plugin/base/security/ForwardingConnectorAccessControl.java (hunk 11 above)
Comment (arhimondr) [reply presto-plugin-toolkit/src/main/java/com/facebook/presto/plugin/base/security/ForwardingConnectorAccessControl.java]: Actually that is mistake. This method has to be moved to the `Add access control checks for SHOW ROLES`

File: presto-spi/src/main/java/com/facebook/presto/spi/security/SystemAccessControl.java (hunk 12 above)
Comment (arhimondr) [reply presto-spi/src/main/java/com/facebook/presto/spi/security/SystemAccessControl.java]: Restored in `Add access control checks for SHOW ROLES`

File: presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java (hunk 14 above)
Comment (arhimondr) [reply presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestRoles.java]: Yeah. Let's keep the name.

**Comment by arhimondr** [comment]
@wenleix Comments addressed
